state-chain/pallets/cf-funding/src/tests.rs

```diff
@@ -1835,3 +1835,139 @@ fn only_governance_can_update_settings() {
+		);
+		assert_eq!(Flip::total_balance_of(&ALICE), AMOUNT, "Total balance to be correct.");
+		assert_ok!(Funding::internal_transfer(
+			OriginTrait::signed(ALICE),
+			BOB,
+			UNRESTRICTED_ADDRESS,
+			AMOUNT_MINUS_FEE.into()
+		));
+		assert_eq!(Flip::total_balance_of(&BOB), AMOUNT_MINUS_FEE, "Total balance to be correct.");
+	});
+}
+
+#[test]
+fn transfer_restricted_funds_internal() {
+	new_test_ext().execute_with(|| {
+		const AMOUNT: u128 = 100;
+		const AMOUNT_MINUS_FEE: u128 = AMOUNT - REDEMPTION_TAX;
+		const RESTRICTED_ADDRESS: EthereumAddress = H160([0x01; 20]);
+
+		assert_ok!(Funding::update_restricted_addresses(
+			RuntimeOrigin::root(),
+			vec![RESTRICTED_ADDRESS],
+			vec![],
+		));
+
+		assert_ok!(Funding::funded(
+			RuntimeOrigin::root(),
+			ALICE,
+			AMOUNT,
+			RESTRICTED_ADDRESS,
+			TX_HASH
+		));
+
+		assert_ok!(Funding::internal_transfer(
+			OriginTrait::signed(ALICE),
+			BOB,
+			RESTRICTED_ADDRESS,
+			AMOUNT_MINUS_FEE.into()
+		));
+
+		assert_eq!(RestrictedBalances::<Test>::get(ALICE).get(&RESTRICTED_ADDRESS), None);
+		assert_eq!(
+			RestrictedBalances::<Test>::get(BOB).get(&RESTRICTED_ADDRESS),
+			Some(&AMOUNT_MINUS_FEE)
+		);
+	});
+}
+
+#[test]
+fn transfer_only_apart_of_the_restricted_funds() {
+	new_test_ext().execute_with(|| {
+		const AMOUNT: u128 = 100;
+		const AMOUNT_MOVE: u128 = AMOUNT / 2;
+		const RESTRICTED_ADDRESS: EthereumAddress = H160([0x01; 20]);
+
+		assert_ok!(Funding::update_restricted_addresses(
+			RuntimeOrigin::root(),
+			vec![RESTRICTED_ADDRESS],
+			vec![],
+		));
+
+		assert_ok!(Funding::funded(
+			RuntimeOrigin::root(),
+			ALICE,
+			AMOUNT,
+			RESTRICTED_ADDRESS,
+			TX_HASH
+		));
+
+		assert_ok!(Funding::internal_transfer(
+			OriginTrait::signed(ALICE),
+			BOB,
+			RESTRICTED_ADDRESS,
+			RedemptionAmount::Exact(AMOUNT_MOVE)
+		));
+
+		// We burn the redemption tax from the restricted balance of the sender.
+		assert_eq!(
+			*RestrictedBalances::<Test>::get(BOB).get(&RESTRICTED_ADDRESS).unwrap(),
+			AMOUNT_MOVE
+		);
+
+		assert_eq!(
+			*RestrictedBalances::<Test>::get(ALICE).get(&RESTRICTED_ADDRESS).unwrap(),
+			AMOUNT - AMOUNT_MOVE - REDEMPTION_TAX
+		);
+	});
+}
+
+#[test]
+fn ensure_bonded_address_condition_holds_during_internal_transfer() {
+	new_test_ext().execute_with(|| {
+		const AMOUNT: u128 = 100;
+		const AMOUNT_MINUS_FEE: u128 = AMOUNT - REDEMPTION_TAX;
+		const RESTRICTED_ADDRESS: EthereumAddress = H160([0x01; 20]);
+
+		assert_ok!(Funding::bind_redeem_address(RuntimeOrigin::signed(ALICE), H160([0x02; 20])));
+		assert_ok!(Funding::bind_redeem_address(RuntimeOrigin::signed(BOB), H160([0x03; 20])));
+
+		assert_ok!(Funding::funded(
+			RuntimeOrigin::root(),
+			ALICE,
+			AMOUNT,
+			RESTRICTED_ADDRESS,
+			TX_HASH
+		));
+
+		assert_noop!(
+			Funding::internal_transfer(
+				OriginTrait::signed(ALICE),
+				BOB,
+				RESTRICTED_ADDRESS,
+				AMOUNT_MINUS_FEE.into()
+			),
+			Error::<Test>::AccountBindingRestrictionViolated
+		);
+	});
+}
```

Comment:
A nice test might be to assert that doing a redeem, then funding using the redeemed amount results in the same outcome as using the `internal_transfer` extrinsic.